In [1]:
#using Pkg
#Pkg.add("Pandas")
using Pandas

In [2]:
dataset = read_csv("ecma_data.csv")

Subject  G1P1  G1P2  G1P3  G1P4  G2P1  G2P2  G2P3  G2P4
0         1     1     3     2     3     1     1     2     3
1         2     2     3     2     2     2     3     3     3
2         3     2     1     1     1     1     2     2     2
3         4     3     1     2     2     1     2     2     3
4         5     3     2     2     1     3     1     1     3
..      ...   ...   ...   ...   ...   ...   ...   ...   ...
75       76     1     2     1     1     3     1     2     3
76       77     1     2     1     1     3     1     2     3
77       78     1     2     1     1     3     1     2     3
78       79     1     2     1     1     3     1     2     3
79       80     1     2     1     1     3     1     2     3

[80 rows x 9 columns]

In [3]:
length(query(dataset, :(G1P1==1&G1P2==2&G1P3==1&G1P4==1)))

62

In Game 1 every level of player other than level 0 chooses the profile of actions `1,2,1,1`.  Of the 80, 62 chose this profile.  This could randomly include some level 0 players.

Estimate the probability $\rho$ that a randomly selected player from this dataset will have level $0$,  To do this, let $\pi$ be the probability that a randomly selected player will choose the profile `1,2,1,1`.  Then $(1-\pi)$ is the probability he or she won't.  Since 
$$\pi = (1-\rho)+\rho (\frac{1}{3})^4$$ we could estimate $\pi$ then calculate an estimate of $\rho$ from that.

Given $\pi$, the probablity of seeing a particular collection of profiles, 62 of which are `1,2,1,1` while 18 aren't is
$$\pi^{62}(1-\pi)^{18}$$

In [4]:
using SymPy
pi,rho = symbols("pi, rho", positive=true)
prob = (pi^(62))*(1-pi)^(18)

 62        18
π  ⋅(1 - π)  

In [5]:
dprob = diff(prob,pi)

      62        17       61        18
- 18⋅π  ⋅(1 - π)   + 62⋅π  ⋅(1 - π)  

In [6]:
solve(dprob,pi)

2-element Vector{Sym}:
 31/40
     1

Using the first element of this vector, the value for $\pi$ that maximizes the probability of seeing this data is $\frac{62}{80}$

Now to estimate $\rho$:

In [7]:
solve((62/80)-(1-rho)-rho*(1/3)^4,rho)

1-element Vector{Sym}:
 0.227812500000000

# Both Games

Assuming only the K-level explanation for behavior there are four profiles that we need to focus on:

Now we calculate profiles:
1. level 1: `1,2,1,1,1,2,1,3`
2. level 2: `1,2,1,1,1,2,2,3`
3. level 3: `1,2,1,1,1,1,2,3`
4. level 4:  `1,2,1,1,3,1,2,3`

Level 4 subject playing as Player 1 in game 2 best replies to the expectation that Player 3 is level 3 who plays 1. The best reply to that is action 3. A level 5 subject as player 1 in Game 2 has the same belief that Player 3 will play 1, so this pattern is repeating.  It is also the same pattern that a fully rational player is supposed to use.  So there is no information in the experiment that would make it possible to tell the difference between levels higher than 3, or whether those players were fully rational (or even Nash).

K level reasoning only explains $\frac{1}{3}$ of the data.

In [8]:
l1 = length(query(dataset, :(G1P1==1&G1P2==2&G1P3==1&G1P4==1&G2P1==1&G2P2==2&G2P3==1&G2P4==3)))
l2 = length(query(dataset, :(G1P1==1&G1P2==2&G1P3==1&G1P4==1&G2P1==1&G2P2==2&G2P3==2&G2P4==3)))
l3 = length(query(dataset, :(G1P1==1&G1P2==2&G1P3==1&G1P4==1&G2P1==1&G2P2==1&G2P3==2&G2P4==3)))
l4 = length(query(dataset, :(G1P1==1&G1P2==2&G1P3==1&G1P4==1&G2P1==3&G2P2==1&G2P3==2&G2P4==3)))
string(l1,"  ",l2,"  ",l3," ",l4)

"7  12  14 13"

What proportion of the players actually use strategies that are consistent with k level reasoning?

In [21]:
(l1+l2+l3+l4)/80

0.575

So the theory explains less than 60% of the data in the experiment. 

Furthermore,  when a player plays one of the profiles listed above, we still aren't sure of their level because the level 0 player could accidentally choose that profile as well. 
Let $\rho_i$ be probability that a randomly drawn player has level $i in {1,2,3,4}$.

Let $\pi_i$ be the probability that a randomly drawn player chooses the profile of actions that would be chosen by a subject of level $i$ and observe that $\pi_i = \rho_i + \rho_0\frac{1}{3^8}$

In [10]:
rho_1,rho_2,rho_3,rho_4,pi_1,pi_2,pi_3,pi_4 = symbols("rho_1,rho_2,rho_3,rho_4,pi_1,pi_2,pi_3,pi_4", positive=true)

(rho_1, rho_2, rho_3, rho_4, pi_1, pi_2, pi_3, pi_4)

As above, let $\pi_i$ be the probability with which profile $i$ is played. Then if $\rho_i$ happened to be the actual probability with which a subject in the pool has level 1, then we could write the probability with which the level 1 action profile `1,2,1,1,1,2,3,3` is played as
$$
\pi_1 = \rho_1+\rho_0(\frac{1}{3})^8
$$
Similary for
$$
\pi_2 = \rho_2+\rho_0(\frac{1}{3})^8
$$
and
$$
\pi_3 = \rho_3+\rho_0(\frac{1}{3})^8
$$
and
$$
\pi_4 = \rho_4+\rho_0(\frac{1}{3})^8
$$

The probability of seeing the data is 
$$
\pi_1^7*\pi_2^{12}*\pi_3^{14}*\pi_4^{13}*(1-\pi_1-\pi_2-\pi_3-\pi_4)^{80-7-12-14}
$$

In [11]:
prob_full = pi_1^7*pi_2^12*pi_3^14*pi_4^13*(1-pi_1-pi_2-pi_3-pi_4)^(80-46)

  7   12   14   13                         34
π₁ ⋅π₂  ⋅π₃  ⋅π₄  ⋅(-π₁ - π₂ - π₃ - π₄ + 1)  

# Method 2 - Numerical

Write down a traditional likelihood function.  It is just the natural log of the `prob_full` symbolic function converted into a function that contains no symbolic stuff.  

In [14]:
llikelihood(x) = (-1)*(7*log(x[1])+12*log(x[2])+14*log(x[3])+13*log(x[4])+(80-7-12-14-13)*log(1-x[1]-x[2]-x[3]-x[4]))


llikelihood (generic function with 1 method)

In [15]:
llikelihood([.3,.3,.1,.2,])

154.32226060755028

In [23]:
import Pkg
install_packages = true
for package in ["BlackBoxOptim", "Distributions", "ForwardDiff", "JSON", "Optim", "Quadrature", "StatsPlots",
    "DotEnv","MySQL","DBInterface","Tables","HTTP", "JSON"]
    if install_packages
        Pkg.add(package)
    end
end
using Optim

    Updating registry at `~/.julia/registries/General.toml`
   Resolving package versions...
   Installed CPUTime ───────────────── v1.0.0
   Installed HypergeometricFunctions ─ v0.3.11
   Installed Rmath_jll ─────────────── v0.3.0+0
   Installed Calculus ──────────────── v0.5.1
   Installed DualNumbers ───────────── v0.6.8
   Installed StatsFuns ─────────────── v1.1.1
   Installed PDMats ────────────────── v0.11.16
   Installed BlackBoxOptim ─────────── v0.6.2
   Installed Rmath ─────────────────── v0.7.0
   Installed DensityInterface ──────── v0.4.0
   Installed Distributions ─────────── v0.25.80
   Installed SpatialIndexing ───────── v0.1.3
   Installed QuadGK ────────────────── v2.7.0
   Installed FillArrays ────────────── v0.13.7
    Updating `~/.julia/environments/v1.8/Project.toml`
  [a134a8b2] + BlackBoxOptim v0.6.2
    Updating `~/.julia/environments/v1.8/Manifest.toml`
  [a134a8b2] + BlackBoxOptim v0.6.2
  [a9c8d775] + CPUTime v1.0.0
  [49dc2e85] + Calculus v0.5.1
  [b429d917

In [24]:
l = [0.05,0.05,0.05,0.05]
u = [0.95,0.95,0.95,0.95]
x0 = [.3,.3,.1,.2]
result = optimize(llikelihood,l,u,x0,Fminbox())

 * Status: success

 * Candidate solution
    Final objective value:     1.169345e+02

 * Found with
    Algorithm:     Fminbox with L-BFGS

 * Convergence measures
    |x - x'|               = 6.05e-10 ≰ 0.0e+00
    |x - x'|/|x'|          = 2.05e-09 ≰ 0.0e+00
    |f(x) - f(x')|         = 0.00e+00 ≤ 0.0e+00
    |f(x) - f(x')|/|f(x')| = 0.00e+00 ≤ 0.0e+00
    |g(x)|                 = 7.04e-09 ≤ 1.0e-08

 * Work counters
    Seconds run:   0  (vs limit Inf)
    Iterations:    4
    f(x) calls:    58
    ∇f(x) calls:   58


In [25]:
parameters = Optim.minimizer(result)

4-element Vector{Float64}:
 0.08750000010402315
 0.1500000000341099
 0.1749999999867283
 0.16250000001169881

Now we have the estimates for the three $\pi$, `.0875`, `.15` and `.175` and `.1625` (all give of take a bit).  As you should probably realize, `.0875` is just $7/80$, so this just illustrates the method.  Now we can find the estimates for the $\rho$.

In [26]:
solve([
        rho_1+(1-rho_1-rho_2-rho_3-rho_4)*(1/3)^8-parameters[1],
        rho_2+(1-rho_1-rho_2-rho_3-rho_4)*(1/3)^8-parameters[2],
        rho_3+(1-rho_1-rho_2-rho_3-rho_4)*(1/3)^8-parameters[3],
        rho_4+(1-rho_1-rho_2-rho_3-rho_4)*(1/3)^8-parameters[4],
        ], rho_1,rho_2,rho_3,rho_4
    )

Dict{Any, Any} with 4 entries:
  rho_4 => 0.162435183784787
  rho_3 => 0.174935183759816
  rho_1 => 0.0874351838771110
  rho_2 => 0.149935183807198

Note that each of the solutions is slightly smaller than the probability with which the corresponding profile appears.  This is intuitively consistent with the idea that playing the profile is incomplete proof that the subject has the corresponding type. 

Finally, we can find the value of the likelihood when we set the $\pi_i$ at their estimated values.

In [27]:
Optim.minimum(result)

116.93447783450424

What does that mean?  You have really just estimated the parameters of a multinomial distribution (the $\pi_i$) using 80 data points.  You can look up the varaince covariance matrix for these estimates on wikipedia. (Or maybe you remember it from your econometrics class).